### ISSUE TITLE
Compress Communication Data

### ISSUE THREAD

**Issue by cabal95** [SparkDevNetwork/Rock#2918]
### Prerequisites
* [X] Put an X between the brackets on this line if you have done all of the following:
    * Can you reproduce the problem on a fresh install or the [demo site](http://rock.rocksolidchurchdemo.com/)?
    * Did you include your Rock version number and [client culture](https://github.com/SparkDevNetwork/Rock/wiki/Environment-and-Diagnostics-Information) setting?
    * Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3ASparkDevNetwork+-repo%3ASparkDevNetwork%2FSlack) to see if your bug or enhancement is already reported?

### Description

> TL;DR = Communication.MediumDataJson can be compressed with a 6:1 ratio. This would save us nearly 2.5GB in database space each year alone. The discussion below is about how this could be achieved and whether this should be a core addition or a plugin.

We have about 1 years worth of Communication data in Rock. Just a hair under actually. We went live in I believe June 2017. Our Communications table is already up to 5GB, 3.7GB of which is in the `MediumDataJson` column. Below is a sum of the size of that column by year. Note that 2015, 2016 and the first half of 2017 is data imported from Arena. Our e-mails were far simpler in Arena because they were more difficult to craft - thus they were much smaller in size.

![image](https://user-images.githubusercontent.com/1119863/38143160-1b1a62a6-33f4-11e8-8eb3-3b28d0bc8e4b.png)

As you can see in 2017, which was not even a full year on Rock, we generated 1.4GB of `MediumDataJson`. In 2018 (as of March 30, 2018) we are at 770MB. That puts us on track to generate 3GB in 2018 - which is about right as 2017 is half that and we were only on Rock for a little more than half a year.

@JimMichael gave me permission to share his numbers as well for a comparison with another church. I'm not sure when he switched to Rock, but based on the numbers I would guess mid 2016:

![image 1](https://user-images.githubusercontent.com/1119863/38143226-62f0df7e-33f4-11e8-94cb-870d9ac700ab.png)

The query to get the above data is here:

```sql
SELECT COUNT([MediumDataJson]) AS [Count], SUM(LEN([MediumDataJson])) AS [Bytes], YEAR(CreatedDateTime) AS [Year] FROM Communication GROUP BY YEAR([CreatedDatetime])
```

### Suggested Plan of Action

The plan would be to compress the `MediumDataJson` data. I ran a test application on all our 2018 data. The test app took just the `MediumDataJson` column, compressed it with System.IO.Compression.GZipStream, then Base-64 encoded it and compared the final size. Actually it made a running total of the uncompressed data size and a running total of the compressed data size and compared at the end.

```
Uncompressed string lengths = 777,043,061 (777MB)
Compressed string lengths = 123,584,900 (123MB)
```

As you can see, that is a 6:1 ratio (actually 6.3:1). Which means we could shrink our existing data from 3.6GB down to around 600MB. Or in terms of our projected per-year numbers, that means a savings of 2.5GB in SQL database space each year.

### Possible Implementations

Note: The following is based on v6 table structure as that is what we are currently running and have been testing on. I believe v7 changed the structure a bit, but I would expect the basic principals to be the same even if some column names have changed.

#### Core

There are two ways this might be implemented. First is an addition to core. As a core addition I would suggest that the data always be compressed before storing to the database and uncompressed as needed, because this would be a "why not" situation. The only issue this might cause is if people are trying to use a DataView (or direct SQL query) that filters on the `MediumDataJson` column, but they would be doing substring search for that so I think it is pretty unlikely.

As a core addition, I think the way the data would be stored would be in a new binary column, which would save another 25% or so of the final data (i.e. that 123MB compressed size would drop to around 85MB) because we would not need to base-64 encode. The existing column could either be left as-is empty or replaced entirely.

The other option would be to keep the base-64 and store it in the existing `MediumDataJson` column encoded as JSON. E.g. `{ "CompressedData": "AA842..." }`. From a quick look through the UI, the only time this data is really used is when viewing Communication History. So that page (possibly others, but this one for sure) would need to be modified to know how to deal with the compressed data - assuming it couldn't be done behind the scenes in the Model, which I hope it could.

If this is implemented in core, a one-shot Job would be created to do the data conversion. Compressing 5-10GB of data during a migration seems like a bad idea.

#### Plugin

The other option is to build a plugin. The basic idea would be the same but it would always store the data base-64 encoded inside the `MediumDataJson`. An attribute would be defined on the Communication model that specifies if it has been compressed or not. Then a job would be created that would run nightly and compress any Communications that are older than xx days (configured by the user, probably default to 90) and not already compressed. It would then obviously mark the attribute to say it's compressed.

This would mean Communication History would work perfectly fine for the first 90 days after a communication is sent. After that it would display no content (though the click tracking, open counts, etc. would still work).

A new block would be created and added to the Communication History page that would detect a compressed communication and display an alert to the user stating the communication has been archived and the content cannot be viewed without unarchiving (i.e. uncompressing). There would also be a button allowing the user to uncompress. When clicked it would uncompress the Communication and redirect the user back to the same page so that it can be viewed. Later when the job runs again it would re-compress the communication.

If implemented as a plugin, I would probably include some other features. For example, one thought we had was that we don't really even need the e-mail body after a certain period of time so even the compressed data could be removed. For us this would likely be a 13-month sliding window. That would give opportunity for staff to say "what did I send out for this event last year, I want to start from that for this year".

### Versions

* **Rock Version:** 6.x
* **Client Culture Setting:** en-US

**Comment by edmistj** [comment]
I added this to the next dev discussion.

**Comment by nairdo** [comment]
@cabal95 I wish I had run this before we updated to 7... Our byte counts are much lower compared to yours and Jim's.  Results from your query:

Count | Bytes | Year
------- | ------ | --------
47196 |	1116 |	2016
84711 |	72 |	2017
16940 |	38 |	2018

I'm pretty certain v7 changes why/where that data is stored now. Looking at the **Interaction** table:

```
SELECT COUNT([InteractionData]) AS [Count], SUM(LEN([InteractionData])) AS [Bytes], YEAR(CreatedDateTime) AS [Year] 
FROM Interaction GROUP BY YEAR([CreatedDatetime])
```

My results...

Count | Bytes | Year
------- | ------ | --------
5972276 |	238316676 |	2017
2473401 |	96135385 |	2018

**Comment by cabal95** [comment]
@nairdo It looks like the "bulk" of the content moved from the `MediumDataJson` to the `Message` column. Can you try this on your v7 and see what comes up?

```sql
SELECT
    COUNT([Message]) AS [Count]
    , SUM(LEN([Message])) AS [Bytes]
    , YEAR(CreatedDateTime) AS [Year]
FROM Communication
GROUP BY YEAR([CreatedDatetime])
```

**Comment by nairdo** [comment]
That results in:

Count	|	Bytes	|	Year
----------|----------------|---------------
1588	|	5690720	|	2007
10139	|	16564657	|	2008
18345	|	114589929	|	2009
22105	|	109296130	|	2010
27328	|	139780879	|	2011
29358	|	191626158	|	2012
35087	|	186834873	|	2013
36766	|	235434257	|	2014
35435	|	238968976	|	2015
46632	|	1074957351	|	2016
84661	|	2417152957	|	2017
22157	|	666184621	|	2018

**Comment by cabal95** [comment]
@nairdo: Looks like your numbers are about the same as Jim's, though your go back a lot further. Dang you guys have been on fancy ChMS systems for a long time. :)

So in 2017 you have about 2.4GB, which should compress down to about 400MB. Looks like all three of us are around the same 650-750MB so far in 2018.

Thanks, helps to verify that this change _should_ help in v7 as well (obviously won't know until I compress it, but I would expect to get similar 6:1 ratio in v7 as well).

**Comment by edmistj** [comment]
@cabal95 we'd like to keep this as is. Now that the bulk of the data is in the Message field it makes it less needed and we'd like to keep the ability to search for items via SQL. 

Thanks though for doing the research and bringing this up!

**Comment by cabal95** [comment]
Okay thanks for letting me know. I'll implement this as a plugin then. 3GB per year is just too much data for us to handle moving forward so we need to shrink that down.

**Comment by nairdo** [comment]
@cabal95 Really? 3GB... how much could 3GB cost in this day-and-age?

**Comment by cabal95** [comment]
To be clear, the concern isn't so much about the raw size of the database, it's about the time it takes to do a database backup. Performing a full DB backup already takes nearly 45 minutes (which we have to do before every upgrade because snapshot is not an option as other databases exist on this SQL server). Our Rock DB is already at 28GB and we have been live for just under one year. When we went live last July our DB was around 12GB. This puts us at a growth rate of over 16GB per year.

Thus we are looking for ways to shrink the size of the database where we can. To us, accessing the raw content of a communication from over 90 days ago is something that almost never happens. So compressing that data (with a convenient "uncompress" button on the page) makes sense for our use case. Especially if that saves us almost 20% of our total growth each year with 0 cost and 0 data loss.

Doing this as a plugin is no problem. We just wanted to be sure we weren't going to write it as a plugin and then be asked to re-write it for core a few months down the road, since the implementation code would be very different that way.

**Comment by edmistj** [comment]
We do consider database size as we write each feature. With the amount of data we collect the database size will increase dramatically over legacy systems. 

Solving this specific use case with a plugin worries me a bit that the intent of the fields changes a bit without core code knowing about the change. I'm not sure we'd feel comfortable having this be a public plugin. If that were the case people may start asking why certain features don't work as expected or show the data the way the documentation shows.

It might be better to use the internal SQL Server compression tools that are transparent to the application. 
https://technet.microsoft.com/en-us/library/dd894051(v=sql.100).aspx

In the end I think we should expect DB sizes to grow much faster than in the past due to more data being captured. The discussion of growth in the Communications tables will look small compared to the growth we'll soon see in the Interactions tables.

But at the end of the day the size of our Rock databases is tiny compared to like systems in other organizations.

**Comment by cabal95** [comment]
Out of curiosity for our own implementation, what features wouldn't work? The only one we could come up with was being able to search communication history by e-mail body content. But that is already impossible to do now as it results in a timeout. With the amount of data we have now, doing a search by message body takes 11 minutes.

**Comment by edmistj** [comment]
It would be hard to answer without first knowing the implementation. But perhaps the viewing of the message in communication history would show the compressed data rather than the actual message content. 

That said I'm more worried that future features/migrations we write would be very difficult if we can be certain of how values are stored in fields. In general changing how data is stored from the original intent of the application developer seems like a bad idea.

Perhaps a better approach would be to write a plugin that wipes out the Message field after a certain period of time and replaces it with a "This message has been archived." with a link to the archived text that is stored in a different location. That at least keeps the Message in plain text and any individual who loads it can understand what happened and take action.

**Comment by cabal95** [comment]
Yes, that was always the plan if implemented as a plugin. In the v6 implementation since it was all JSON data it could have been stored in the JSON data under a different key. In v7 with a dedicated column for the message body it would have to be an Attribute to store the compressed data. and the Message column either blanked out or a placeholder string.

**Comment by edmistj** [comment]
Sorry that was not clear. I read it as "An attribute would be defined on the Communication model that specifies if it has been compressed or not. " which made it sound like the data would be changed in the current fields with a flag indicating it had been changed.

Reading through the solution description above I would hope you'd be able to architect it in a way that no extra blocks would need to be added to the communication history page. Having the message body contain a link to a new page to view the original message would seem to be elegant.

I would though encourage some research into the compression features in SQL Server. They may meet your need without any code.

**Comment by cabal95** [comment]
Sorry I wasn't more clear on that. Most of the original message was written before I realized exactly how v7 changed the way it stores the data.

We might be able to do the link method to link to a page that does the decompress and then redirects back to the communication history page. I'll have to play with it and see if that works well. Thanks for the thought.

TL;DR below = SQL Server compression does  not work on the Communication table because the row data is more than 8,000 bytes.

Unfortunately it doesn't look like the SQL Server compression will work in this case. Apparently it only compresses in-row data (even when using PAGE compression, oddly enough). VARCHAR(MAX) and related column types can only be stored in-row if their length is < 8,000, actually it might be the entire row must be < 8,000. Not sure on the implementation specifics. But anyway, the end result is that SQL Server won't compress the bulk of the data in these tables.

I used the built-in stored procedure to estimate space savings on a v7 Communication table with 752MB of space used (This is a copy of our data as of October 2016 when we were still in process of migration, so it contains only our migrated data, thus the small size). It estimated the resulting compressed table would take up 748MB of space, so it that seems to correlate with the information I found online about large columns not being compressible via standard compression methods.

So I think the only way to do it "natively" in SQL Server would be with a FILESTREAM that stores the data on a compressed volume. But that would, as far as I know, require changing the structure of the Rock tables.

https://dba.stackexchange.com/questions/169254/attempting-to-compress-an-existing-table-results-in-no-change-to-the-table-size

It might be worth using on other tables such as Interactions. The "estimate compression savings" sproc on my dev box shows an estimated 50% savings there. I'll do some more research on this and post an STB article or something if I find some good space savings possibilities for using the built-in compression settings.
